File: types/product.go
package types

import (
	"encoding/json"
	"strings"
)

// FIXME update ALL ID fields to be int64 to avoid implicit type conversion by Postgres
// The below example fixes this issue/overhead while still returning ID fields as string to UI
// ID int64 `json:"id,string"` // Serializes as string in JSON
type Product struct {
	ID          string          `json:"id"`
	Name        string          `json:"name"`
	Price       int64           `json:"price"`
	Details     json.RawMessage `json:"details"`
	Summary     string          `json:"summary"`
	Description *string         `json:"description,omitempty"`
	Images      []Image         `json:"images"`
	Category    *Category       `json:"category"`
	TaxCode     *string         `json:"tax_code,omitempty"`
	Inventory   int             `json:"inventory"`
	Featured    bool            `json:"featured"`
	CartLimit   *int            `json:"cart_limit,omitempty"`
	CreatedAt   string          `json:"created_at"`
	UpdatedAt   string          `json:"updated_at"`
}

type ImageType string

const (
	Thumbnail ImageType = "thumbnail"
	Gallery   ImageType = "gallery"
	Hero      ImageType = "hero"
)

func ParseImageType(s string) ImageType {
	switch strings.ToLower(s) {
	case "hero":
		return Hero
	case "thumbnail":
		return Thumbnail
	case "gallery":
		return Gallery
	default:
		return Gallery
	}
}

type Image struct {
	ID        string    `json:"id"`
	ProductID string    `json:"product_id"`
	URL       string    `json:"url"`
	Type      ImageType `json:"type"`
	AltText   *string   `json:"alt_text,omitempty"` // FIXME convert pointer to string
	Source    string    `json:"source"`
}

type ProductFilter struct {
	SortBy     SortBy
	InStock    bool
	Featured   bool
	Page       int
	Limit      int
	Categories []string // category slugs
}

type SortBy string

const (
	SortByPrice      SortBy = "price"
	SortByPopularity SortBy = "total_sold"
	SortByNewest     SortBy = "created_at"
	SortByDefault    SortBy = "created_at"
)

func ParseSortBy(sortBy string) SortBy {
	switch sortBy {
	case "price":
		return SortByPrice
	case "popularity":
		return SortByPopularity
	case "newest":
		return SortByNewest
	default:
		return SortByDefault
	}
}

File: web/src/types/product.ts
import type { Category } from './category'

export interface Product {
  id: string
  name: string
  price: number
  summary: string
  description: string
  details: Record<string, unknown>
  images: Image[]
  category?: Category
  tax_code?: string
  inventory: number
  created_at?: string
  updated_at?: string
}

export type SortBy = 'price' | 'popularity' | 'newest'

export interface ProductFilters {
  categories?: string[]
  sortBy?: SortBy
  inStock?: boolean
  page?: number
  limit?: number
}

export interface CreateProductRequest {
  name: string
  price: number
  details: Record<string, unknown>
  description: string
  tax_code?: string
}

type RequireOnly<T, K extends keyof T> = Partial<T> & Pick<T, K>

export type UpdateProductRequest = RequireOnly<Product, 'id'>

export type ImageType = 'hero' | 'thumbnail' | 'gallery'

export interface Image {
  id: string
  product_id?: string // FIXME pretty sure this never is used/exists
  url: string
  type: ImageType
  alt_text?: string | null
}

export interface CartItem {
  quantity: number
  product: Product
  unit_price: number
}
